pkg/controllers/tagging/tagging_controller.go

```diff
@@ -326,8 +349,14 @@ func (tc *Controller) enqueueNode(node *v1.Node, action func(node *v1.Node) erro
 		requeuingCount: 0,
 		enqueueTime:    time.Now(),
 	}
-	tc.workqueue.Add(item)
-	klog.Infof("Added %s to the workqueue", item)
+
+	if tc.rateLimitEnabled {
+		tc.workqueue.AddRateLimited(item)
+		klog.Infof("Added %s to the workqueue (rate-limited)", item)
+	} else {
+		tc.workqueue.Add(item)
```

Comment:
In case of errors, we always add the item under a rate limit regardless of whether or not these new flags are enabled - https://github.com/kubernetes/cloud-provider-aws/blob/master/pkg/controllers/tagging/tagging_controller.go#L215.

Reply:
So here you're choosing between `Add` and `AddRateLimited` based on whether `rateLimitEnabled`. Wondering why behavior of adding new items is depending on the queue definition.

Also something to consider - rate-limiting at the workqueue-level (instead of the ec2 client directly) can be riskier because there could be multiple points in the controller code where you enqueue items into the queue (currently you have two - new node, retried node - potentially more in future). Each of them needs to be careful about the way it's adding items to the queue (i.e `Add` vs `AddRateLimited`). If you did the same at client-level, it's more fool-proof because the setting is local to the client definition and is **the only** exit point for any call to ec2.

Reply:
You're correct and there was a deliberate reason to add this rate limiting on the queue at the moment. It is because the EC2 client today isn't controller specific (there's one at the CCM level which is used by all controllers). 

As a result, adding a rate limit there would mean that all controllers would get throttled which might not be something that we want (e.g. we may want the node controller to not be throttled but tagging controller to be throttled). Adding this rate limit on the queue level gives us this option.